### DISCUSSION TITLE
How does Codegen work?

### DISCUSSION THREAD

**Discussion by osdiab** [chakra-ui/panda#845] [Q&A] [answered]
### Subject

Codegen

### Description

The docs don't actually say anywhere how the `codegen` command works, just that you should refer to it for details (and there are none basically). So I was surprised that just having the `codegen` `prepare` script somehow instruments the NextJS server to actually watch for style changes and panda to regenerate code automatically. Very magical to not have to run `panda --watch` or anything. Would love for this to be clearer how it works.

**Comment by astahmer** [comment, marked as answer]
I don't think the `prepare` script helps with watching files, but there is some file watching in the `postcss` plugin which you are probably (?) using 

it's done here
https://github.com/chakra-ui/panda/blob/0a11c5965a8f3100b5cbe982f812bb37d2e68c32/packages/node/src/builder.ts#L93
tl;dr is: `postcss plugin` uses the `Builder` which finds a list of file dependencies for the `panda.config.ts` (from what was imported, recursively) and then checks that any of those config dependencies has changed [whenever the postcss plugin notifies the builder](https://github.com/chakra-ui/panda/blob/fix/parser-imports-with-baseurl/packages/postcss/src/index.ts#L26) that something has changed

**Reply by osdiab** [reply]
nice - that makes more sense. @segunadebayo if info like this can be included in some advanced part of the docs that would be super helpful! i'm a fan of understanding how my tools work so when things get too "magic" i get wary.
